packages/legend-studio/src/stores/editor-state/element-editor-state/connection/DatabaseBuilderState.ts

```diff
@@ -0,0 +1,602 @@
+}
+
+export class DatabaseBuilderState {
+  editorStore: EditorStore;
+  connection: RelationalDatabaseConnection;
+  modal = false;
+  databaseGrammar = '';
+  isBuildingDatabase = false;
+  isSavingDatabase = false;
+  targetDatabasePath: string;
+  WILDCARD = '%';
+  treeData?: DatabaseBuilderTreeData;
+
+  constructor(
+    editorStore: EditorStore,
+    connection: RelationalDatabaseConnection,
+  ) {
+    makeObservable(this, {
+      modal: observable,
+      targetDatabasePath: observable,
+      setTargetDatabasePath: action,
+      setModal: action,
+      setDatabaseGrammar: action,
+      setTreeData: action,
+      isBuildingDatabase: observable,
+      databaseGrammar: observable,
+      isSavingDatabase: observable,
+      treeData: observable,
+    });
+
+    this.connection = connection;
+    this.editorStore = editorStore;
+    this.targetDatabasePath = this.currentDatabase?.path ?? 'store::MyDatabase';
+  }
+
+  setModal(val: boolean): void {
+    this.modal = val;
+  }
+
+  setTreeData(builderTreeData?: DatabaseBuilderTreeData): void {
+    this.treeData = builderTreeData;
+  }
+
+  setDatabaseGrammar(val: string): void {
+    this.databaseGrammar = val;
+  }
+
+  setTargetDatabasePath(val: string): void {
+    this.targetDatabasePath = val;
+  }
+
+  // Tree Operations
+  onNodeSelect = flow(function* (
+    this: DatabaseBuilderState,
+    node: DatabaseBuilderTreeNodeData,
+    treeData: DatabaseBuilderTreeData,
+  ) {
+    if (
+      node instanceof SchemaDatabaseBuilderTreeNodeData &&
+      !node.childrenIds
+    ) {
+      yield this.fetchSchemaTableDefinitions(node, treeData);
+    } else if (
+      node instanceof TableDatabaseBuilderTreeNodeData &&
+      !node.childrenIds
+    ) {
+      yield this.fetchTableMetadata(node, treeData);
+    }
+    node.isOpen = !node.isOpen;
+    this.setTreeData({ ...treeData });
+  });
+
+  getChildNodes(
+    node: DatabaseBuilderTreeNodeData,
+    treeData: DatabaseBuilderTreeData,
+  ): DatabaseBuilderTreeNodeData[] | undefined {
+    return node.childrenIds
+      ?.map((n) => treeData.nodes.get(n))
+      .filter(isNonNullable);
+  }
+
+  toggleCheckedNode(
+    node: DatabaseBuilderTreeNodeData,
+    treeData: DatabaseBuilderTreeData,
+  ): void {
+    node.isChecked = !node.isChecked;
+    if (node instanceof SchemaDatabaseBuilderTreeNodeData) {
+      this.getChildNodes(node, treeData)?.forEach((n) => {
+        n.isChecked = node.isChecked;
+      });
+    } else if (node instanceof TableDatabaseBuilderTreeNodeData) {
+      if (node.parentId) {
+        const parent = treeData.nodes.get(node.parentId);
+        if (
+          parent &&
+          this.getChildNodes(parent, treeData)?.every(
+            (e) => e.isChecked === node.isChecked,
+          )
+        ) {
+          parent.isChecked = node.isChecked;
+        }
+      }
+    }
+    this.setTreeData({ ...treeData });
+  }
+
+  buildNonEnrichedDbBuilderInput(schema?: string): DatabaseBuilderInput {
```

Comment:
When is this `enriched` then?

Reply:
when you open the table node you enrich it to get the columns and their type